WordPress/openverse-frontend
# Add progressive TypeScript compilation

## Fixes


Fixes #175 by @sarayourfriend

## Description


Replaces `jsconfig.json` with a `tsconfig.json` that mirrors Gutenberg's with some changes to accomodate not being a composite project.

Currently this only checks a single file, basically in an effort to minimize the size of this PR while also still proving that it does work as intended. I had to make a small update to the type declarations in this file to get the compiler to pass. Hooray for type feedback! 🎉 

Future work will include slowly adding each file in `src/composables` to the `include` array. We can also start running the type checker on other parts of the codebase like the services 🙂 I would propose doing this in small isolated bits and pieces to avoid having to make too many changes in a single PR to address the inevitable error here and there.

## Testing Instructions

Checkout the PR and run `npm run types`. Everything should come back without an error. Next, update `use-event-listener-outside.js` to include a type error. For example, add the following somewhere in the file:

```js
/** @type {string} */
const foo = null
```

This is a type error becuase you can't assign `null` to a `string`. Your IDE should pick this up and give you some squiggly lines or a warning (however your IDE normally picks up compilation errors). You should now also be able to run `npm run tsc` and have it report back the error to you.

Go ahead and try to commit that (you'll have to disable the eslint `no-unused-vars` rule temporarily to be able to run the pre-commit) and pre-commit should fail on the `tsc` step.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
